site/app/controllers/pdf/PDFController.php

```diff
@@ -251,4 +304,155 @@ public function showGraderPDFEmbedded(string $gradeable_id) {
+        $filename = $_POST['filename'] ?? null;
+        // Explicit boolean parsing for is_anon
+        $is_anon = filter_var($_POST['is_anon'] ?? false, FILTER_VALIDATE_BOOL);
+        $filename = html_entity_decode($filename);
+        $file_path = urldecode($_POST['file_path']);
+        $real_path = $is_anon ? "" : $file_path;
+        $anon_path = $is_anon ? $file_path : "";
+
+        if ($is_anon) {
+            $id = $this->core->getQueries()->getSubmitterIdFromAnonId($id, $gradeable_id);
+            $real_path = $this->getPath($file_path, $id, true);
+            if (!file_exists($real_path)) {
+                $this->core->getOutput()->renderJsonFail('The image file could not be found');
+                return;
+            }
+        }
+        else {
+            $anon_path = $this->getPath($file_path, $gradeable_id, false);
+        }
+
+        $gradeable = $this->tryGetGradeable($gradeable_id);
+        if ($gradeable === false) {
+            $this->core->getOutput()->renderJsonFail('Could not get gradeable');
+            return;
+        }
+
+        $graded_gradeable = $this->tryGetGradedGradeable($gradeable, $id);
+        if ($graded_gradeable === false) {
+            $this->core->getOutput()->renderJsonFail('Could not get graded gradeable');
+            return;
+        }
+
+        if (!$this->core->getAccess()->canI("grading.electronic.grade", ["gradeable" => $gradeable, "graded_gradeable" => $graded_gradeable])) {
+            $this->core->getOutput()->renderJsonFail('You do not have permission to grade this student');
+            return;
+        }
+
+        // We've already verified that we can grade this assignment.  We just check to see if this a peer grader
+        // to determine if we should show a button to download the image.
+        $is_peer_grader = $this->core->getUser()->getGroup() === User::GROUP_STUDENT && $gradeable->hasPeerComponent();
+
+        $active_version = $graded_gradeable->getAutoGradedGradeable()->getActiveVersion();
+        $annotation_dir = FileUtils::joinPaths($this->core->getConfig()->getCoursePath(), 'annotations', $gradeable_id, $id, $active_version);
+        $annotation_jsons = [];
+        $file_path_md5 = md5($file_path);
+        if (is_dir($annotation_dir)) {
+            $dir_iter = new \FilesystemIterator($annotation_dir);
+            foreach ($dir_iter as $annotation_file) {
+                if (explode('_', $annotation_file->getFilename())[0] === $file_path_md5) {
+                    $file_contents = file_get_contents($annotation_file->getPathname());
+                    $annotation_decoded = json_decode($file_contents, true);
+                    if ($annotation_decoded !== null) {
+                        $grader_id = $annotation_decoded["grader_id"];
+                        $annotation_jsons[$grader_id] = json_encode($annotation_decoded['annotations']);
+                    }
+                }
+            }
+        }
```

Comment:
Can you add comments or factor some of this out into a separate function so it's clear what's going on here?